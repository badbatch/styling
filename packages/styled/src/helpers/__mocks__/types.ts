export interface MockBuildBaseSelector {
  _setSelector: (selector: string) => void;
  default: (componentName: string, prefix?: string) => string;
}
