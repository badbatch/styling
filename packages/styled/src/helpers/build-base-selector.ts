import { kebabCase } from "lodash";
import shortid from "shortid";

export default function buildBaseSelector(componentName: string, prefix?: string) {
  return `${prefix ? `${prefix}-` : ""}${kebabCase(componentName)}-${shortid.generate()}`;
}
