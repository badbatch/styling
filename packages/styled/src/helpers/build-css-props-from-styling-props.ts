import { PlainObject } from "@styling/types";
import { kebabCase } from "lodash";
import { StylingCSSVariables } from "../types";

export default function buildCSSPropsFromStylingProps(combo: string[], cssVariableProps: StylingCSSVariables) {
  const props: PlainObject = {};

  cssVariableProps.forEach(([key, value]) => {
    props[key] = `var(--${kebabCase(key)}${value ? `, ${value}` : ""})`;
  });

  combo.forEach(prop => {
    const split = prop.split("::");

    if (split.length === 2) {
      props[split[0]] = split[1];
    } else {
      props[split[0]] = true;
    }
  });

  return props;
}
