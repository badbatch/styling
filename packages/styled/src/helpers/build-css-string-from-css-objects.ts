import autoprefixer from "autoprefixer";
import postcss from "postcss";
import postcssJs from "postcss-js";
import { PropKeyComboCSS } from "../types";

export default function buildCSSStringFromCSSObjects(propKeyComboCSS: PropKeyComboCSS) {
  let css = "";
  const keys = Object.keys(propKeyComboCSS);

  for (const key of keys) {
    const result = postcss([autoprefixer]).process(
      { [`.${propKeyComboCSS[key].selector}`]: propKeyComboCSS[key].css },
      {
        from: undefined,
        parser: postcssJs.parse,
      },
    );

    css += `${result.css}\n`;
  }

  return css;
}
