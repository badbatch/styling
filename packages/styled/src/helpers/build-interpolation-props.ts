import { CSSVariablePropList, PlainObject } from "@styling/types";
import { kebabCase } from "lodash";

export default function buildInterpolationProps(propKeyCombo: string[], cssVariablePropList: CSSVariablePropList) {
  const props: PlainObject = {};

  cssVariablePropList.forEach(([key, value]) => {
    props[key] = `var(--${kebabCase(key)}${value ? `, ${value}` : ""})`;
  });

  propKeyCombo.forEach(prop => {
    const split = prop.split("::");

    if (split.length === 2) {
      props[split[0]] = split[1];
    } else {
      props[split[0]] = true;
    }
  });

  return props;
}
