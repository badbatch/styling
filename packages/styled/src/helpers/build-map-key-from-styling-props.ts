import { kebabCase } from "lodash";

export default function buildMapKeyFromStylingProps(propNameCombo: string[]) {
  return propNameCombo
    .map(propName => kebabCase(propName))
    .sort()
    .join("--");
}
