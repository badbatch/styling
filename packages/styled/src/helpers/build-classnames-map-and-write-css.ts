import { error, info } from "@styling/helpers";
import { CSSVariablePropList, Interpolation, StringObject } from "@styling/types";
import { Metadata } from "../types";
import buildCSSObjects from "./build-css-objects";
import buildCSSStringFromCSSObjects from "./build-css-string-from-css-objects";
import loadStylingConfig from "./load-styling-config";
import writeCSS from "./write-css";

export default function buildClassNamesMapAndWriteCSS(
  propKeyCombos: string[][],
  cssVariablePropList: CSSVariablePropList,
  interpolations: Interpolation[],
  { componentName, sourceDir }: Metadata,
) {
  const config = loadStylingConfig({ componentName, sourceDir });

  info("Building css objects");
  const propKeyComboCSS = buildCSSObjects(propKeyCombos, cssVariablePropList, interpolations, componentName, config);
  const { outputPath } = config;

  info("Generating css from css objects");
  const css = buildCSSStringFromCSSObjects(propKeyComboCSS);

  info(`Writing css to ${outputPath}`, css);

  try {
    writeCSS(css, outputPath);
  } catch (e) {
    error("Writing css failed", e);
  }

  return {
    propsToClassNamesMap: Object.keys(propKeyComboCSS).reduce((map: StringObject, key) => {
      const { selector } = propKeyComboCSS[key];
      map[key] = selector;
      return map;
    }, {}),
    relevantPropKeys: Object.keys(propKeyComboCSS).reduce((set, key) => {
      const { keyCombo } = propKeyComboCSS[key];
      return [...new Set([...set, ...keyCombo])];
    }, [] as string[]),
  };
}
