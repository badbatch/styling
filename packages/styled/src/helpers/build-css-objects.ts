import { buildMapKeyFromPropKeyCombo } from "@styling/helpers";
import { CSSVariablePropList, Interpolation, StylingConfig } from "@styling/types";
import { PropKeyComboCSS } from "../types";
import buildBaseSelector from "./build-base-selector";
import buildInterpolationProps from "./build-interpolation-props";
import buildSelectorFromPropKeyCombo from "./build-selector-from-prop-key-combo";
import collateCSS from "./collate-css";
import dedupeCSS from "./dedupe-css";

export default function buildCSSObjects(
  propKeyCombos: string[][],
  cssVariablePropList: CSSVariablePropList,
  interpolations: Interpolation[],
  componentName: string,
  { selectorPrefix, theme }: StylingConfig,
) {
  const baseSelector = buildBaseSelector(componentName, selectorPrefix);
  const baseCSS = collateCSS(interpolations, buildInterpolationProps([], cssVariablePropList), theme);

  const propKeyComboCSS: PropKeyComboCSS = {};

  if (Object.keys(baseCSS).length) {
    propKeyComboCSS.base = {
      css: baseCSS,
      keyCombo: [],
      selector: baseSelector,
    };
  }

  const sortedKeyCombos = propKeyCombos.sort((a, b) => {
    return a.length - b.length;
  });

  sortedKeyCombos.forEach(keyCombo => {
    const comboCSS = dedupeCSS(
      collateCSS(interpolations, buildInterpolationProps(keyCombo, cssVariablePropList), theme),
      keyCombo,
      propKeyComboCSS,
    );

    if (Object.keys(comboCSS).length) {
      propKeyComboCSS[buildMapKeyFromPropKeyCombo(keyCombo)] = {
        css: comboCSS,
        keyCombo,
        selector: buildSelectorFromPropKeyCombo(baseSelector, keyCombo),
      };
    }
  });

  return propKeyComboCSS;
}
