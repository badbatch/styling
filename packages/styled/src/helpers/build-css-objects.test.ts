import { generatePropKeyCombos } from "@styling/helpers";
import { PlainObject, PropListWithoutCSSVariables } from "@styling/types";
import { getInterpolations } from "../__tests__/helpers";
import { MockBuildBaseSelector } from "../types";
import buildCSSObjects from "./build-css-objects";

jest.mock("./build-base-selector");
const buildBaseSelector = require("./build-base-selector") as MockBuildBaseSelector;

describe("buildCSSObjects", () => {
  buildBaseSelector._setSelector("container-eHB-OhMi");

  const propList = [
    "block",
    "disabled",
    "error",
    "externalText",
    "inverse",
    "noText",
    ["size", ["xs", "sm", "md"]],
    "stretch",
    ["variant", ["primary", "secondary", "link"]],
  ];

  const baseInterpolations = getInterpolations`
    -webkit-tap-highlight-color: transparent;
    box-sizing: border-box;
    outline: none;
    position: relative;
  `;

  const displayInterpolations = getInterpolations`
    align-items: center;
    display: table-cell;
    display: flex;
    justify-content: center;
    text-align: center;
    vertical-align: middle;
`;

  function getBorderRadius({ noText, variant }: PlainObject) {
    if (variant === "link") return "0";
    return noText ? "50%" : "24px";
  }

  function getInnerContainerJustifyContent({ stretch }: PlainObject) {
    return stretch && "justify-content: space-between;";
  }

  const interpolations = getInterpolations`
    ${baseInterpolations};
    border-radius: ${getBorderRadius};
    height: 100%;
    white-space: nowrap;
    width: 100%;
    ${displayInterpolations};
    ${getInnerContainerJustifyContent};
  `;

  it("SHOULD return the correct selector css objects", () => {
    const propNameComboCSS = buildCSSObjects(
      generatePropKeyCombos(propList as PropListWithoutCSSVariables),
      [],
      interpolations,
      "Container",
      {
        outputPath: "",
        selectorPrefix: "button",
        theme: {},
      },
    );

    expect(propNameComboCSS).toMatchSnapshot();
  });
});
