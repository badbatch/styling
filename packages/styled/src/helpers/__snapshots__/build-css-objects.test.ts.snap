// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildCSSObjects SHOULD return the correct selector css objects 1`] = `
Object {
  "base": Object {
    "css": Object {
      "WebkitTapHighlightColor": "transparent",
      "alignItems": "center",
      "borderRadius": "24px",
      "boxSizing": "border-box",
      "display": "flex",
      "height": "100%",
      "justifyContent": "center",
      "outline": "none",
      "position": "relative",
      "textAlign": "center",
      "verticalAlign": "middle",
      "whiteSpace": "nowrap",
      "width": "100%",
    },
    "keyCombo": Array [],
    "selector": "container-eHB-OhMi",
  },
  "no-text": Object {
    "css": Object {
      "borderRadius": "50%",
    },
    "keyCombo": Array [
      "noText",
    ],
    "selector": "container-eHB-OhMi--no-text",
  },
  "stretch": Object {
    "css": Object {
      "justifyContent": "space-between",
    },
    "keyCombo": Array [
      "stretch",
    ],
    "selector": "container-eHB-OhMi--stretch",
  },
  "variant-link": Object {
    "css": Object {
      "borderRadius": "0",
    },
    "keyCombo": Array [
      "variant::link",
    ],
    "selector": "container-eHB-OhMi--variant-link",
  },
}
`;
